File: frontend/src/types/stats.ts
import type { StatusCounts } from "@/types/book";

export interface GenreCount {
  genre: string;
  count: number;
}

export interface DecadeCount {
  decade: string;
  count: number;
}

export interface ReadingPeriodCounts {
  books: number;
  pages: number;
}

export interface ReadingAverages {
  books_per_day: number;
  books_per_week: number;
  books_per_month: number;
  books_per_year: number;
  pages_per_day: number;
  pages_per_week: number;
  pages_per_month: number;
  pages_per_year: number;
}

export interface Stats {
  total_books: number;
  status_counts: StatusCounts;
  total_pages: number;
  avg_publication_year: number | null;
  most_common_author: string | null;
  most_common_genre: string | null;
  genre_counts: GenreCount[];
  decade_counts: DecadeCount[];
  pages_read_total: number;
  reading_this_week: ReadingPeriodCounts;
  reading_this_month: ReadingPeriodCounts;
  reading_this_year: ReadingPeriodCounts;
  reading_averages: ReadingAverages | null;
}

File: backend/app/schemas/stats.py
from pydantic import BaseModel

from app.schemas.book import StatusCounts


class GenreCount(BaseModel):
    genre: str
    count: int


class DecadeCount(BaseModel):
    decade: str
    count: int


class ReadingPeriodCounts(BaseModel):
    books: int
    pages: int


class ReadingAverages(BaseModel):
    books_per_day: float
    books_per_week: float
    books_per_month: float
    books_per_year: float
    pages_per_day: float
    pages_per_week: float
    pages_per_month: float
    pages_per_year: float


class StatsOut(BaseModel):
    total_books: int
    status_counts: StatusCounts
    total_pages: int
    avg_publication_year: int | None
    most_common_author: str | None
    most_common_genre: str | None
    genre_counts: list[GenreCount]
    decade_counts: list[DecadeCount]
    pages_read_total: int
    reading_this_week: ReadingPeriodCounts
    reading_this_month: ReadingPeriodCounts
    reading_this_year: ReadingPeriodCounts
    reading_averages: ReadingAverages | None
